### PR TITLE
Ribbon

### PR DESCRIPTION
**Pull request by vgromfeld**
## What?
Adding the `Ribbon` control to the toolkit.

See https://github.com/CommunityToolkit/Labs-Windows/discussions/544 and https://github.com/CommunityToolkit/Labs-Windows/issues/545 for more details.

## How?
I'm adding the following components to the toolkit:

- **Ribbon** the ribbon control.
- **RibbonGroup** an always displayed group container.
- **RibbonCollapsibleGroup** a group container which can be collapsed based on a priority order if there is not enough space available to fully display the group.
- **RibbonPanel** a custom panel for the Ribbon to automatically expand or collapse the RibbonCollapsibleGroups
- **RibbonStyle.xaml** all the custom styles for the **Ribbon** components.

### Usage
```xml
<Ribbon>
   <RibbonGroup Label="Selection" />
    <RibbonCollapsibleGroup Label="Edit" Style="{StaticResource RibbonLeftCollapsibleGroupStyle}" Priority="1">
        <!-- group content -->
    </RibbonCollapsibleGroup>
    <RibbonCollapsibleGroup Label="Shapes" Style="{StaticResource RibbonLeftCollapsibleGroupStyle}" Priority="2">
        <!-- group content -->
    </RibbonCollapsibleGroup>
</Ribbon>
```

### Result
image
image

Ribbon

### CODE REVIEW

**Comment by mdtauk** [comment]
What would be the performance and usability implications to combining this Ribbon with a SelectorBar control, to give you the same functionality of an Office like Ribbon and switching between ribbon sets?

image

**Comment by mdtauk** [comment]
> It should be easy to combine the two. The `SelectorBar` would provide the "tabs" and the `Ribbon` the buttons.

So you would use the Selected **SelectorBarItem** to load in a **Ribbon**?
The **Ribbon** itself contains **RibbonGroup**s which contain **RibbonCollapsibleGroup**s 

I guess the Collection of Ribbons should be virtualised, so when switching to the Active Ribbon, doesn't keep them loaded?  Or should all Ribbons be loaded by this theoretical Control, to enable smooth switching?

I guess this is the performance aspect I am asking about

**Comment by Tartarusome** [comment]
> > It should be easy to combine the two. The  would provide the "tabs" and the  the buttons.`SelectorBar``Ribbon`
> 
> So you would use the Selected **SelectorBarItem** to load in a **Ribbon**? The **Ribbon** itself contains **RibbonGroup**s which contain **RibbonCollapsibleGroup**s
> 
> I guess the Collection of Ribbons should be virtualised, so when switching to the Active Ribbon, doesn't keep them loaded? Or should all Ribbons be loaded by this theoretical Control, to enable smooth switching?
> 
> I guess this is the performance aspect I am asking about


i suppose no,  the CPP file in Windows App Sdk source code dosent have a disppose method，but icon and text block resources should be autumatically disposed by XAML frame. I didnt take a test.
If you want a custom control to be visualized, you should make it inherit from IDisposable besides Control and write the dispose method.

**Comment by michael-hawker** [comment]
Some build errors:

```
"D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\CommunityToolkit.WinUI.Controls.Ribbon.csproj" (default target) (17:93) ->
  D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\Ribbon.cs(22,21): error CS0260: Missing partial modifier on declaration of type 'Ribbon'; another partial declaration of this type exists [D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\CommunityToolkit.WinUI.Controls.Ribbon.csproj::TargetFramework=net7.0-android33.0]
  D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\RibbonCollapsibleGroup.cs(19,14): error CS0260: Missing partial modifier on declaration of type 'RibbonCollapsibleGroup'; another partial declaration of this type exists [D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\CommunityToolkit.WinUI.Controls.Ribbon.csproj::TargetFramework=net7.0-android33.0]
  D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\RibbonGroup.cs(12,14): error CS0260: Missing partial modifier on declaration of type 'RibbonGroup'; another partial declaration of this type exists [D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\CommunityToolkit.WinUI.Controls.Ribbon.csproj::TargetFramework=net7.0-android33.0]
  D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\RibbonPanel.cs(11,23): error CS0260: Missing partial modifier on declaration of type 'RibbonPanel'; another partial declaration of this type exists [D:\a\Labs-Windows\Labs-Windows\components\Ribbon\src\CommunityToolkit.WinUI.Controls.Ribbon.csproj::TargetFramework=net7.0-android33.0]
```

Look similar between UWP/WinUI3, looks like something going on with Uno. I think Uno requires all UI classes to be partial for the generators to work, so probably just that?

File: components/Ribbon/src/Ribbon.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+    public Ribbon()
+    {
+        DefaultStyleKey = typeof(Ribbon);
+
+        _items = [];
+        _items.CollectionChanged += OnItemsCollectionChanged;
+    }
+
+    public ICollection<UIElement> Items => _items;
```
Comment (michael-hawker) [inline components/Ribbon/src/Ribbon.cs:66]: Wonder if this should be `IList`, that may not matter with the setup you have for this here, but I know we've had other XAML related issues for other collection type controls we've had in the past as some internal XAML pieces expect IList over other collection types.

**Review by michael-hawker — COMMENTED** [review]
Few comments, nothing too critical I don't think, but some suggestions/questions.

File: components/Ribbon/samples/RibbonCustomSample.xaml [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        <ResourceDictionary Source="ms-appx:///CommunityToolkit.WinUI.Controls.Ribbon/RibbonStyle.xaml" />
+    </Page.Resources>
+
+    <StackPanel Spacing="16">
+        <StackPanel.Resources>
+            <Style BasedOn="{StaticResource DefaultAppBarButtonStyle}"
+                   TargetType="AppBarButton">
+                <Setter Property="LabelPosition" Value="Collapsed" />
+                <Setter Property="Width" Value="48" />
+                <Setter Property="Margin" Value="8" />
```
Comment (0x5bfa) [inline components/Ribbon/samples/RibbonCustomSample.xaml:25]: Isn’t 8 a bit bigger?
As a sample we might want to provide devs how this control should be used and should look like. I assume many devs might want to make it look like Paint app so making it alike is probably good idea.

File: components/Ribbon/src/RibbonStyle.xaml [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+                            <RowDefinition Height="Auto" />
+                        </Grid.RowDefinitions>
+
+                        <ContentControl x:Name="VisibleContentContainer"
+                                        HorizontalAlignment="Center"
+                                        VerticalAlignment="Center"
+                                        IsTabStop="False"
+                                        Visibility="Collapsed" />
+
+                        <Button x:Name="CollapsedButton"
```
Comment (0x5bfa) [inline components/Ribbon/src/RibbonStyle.xaml:112]: Although this might be already discussed and this is out of scope in this initial PR, might be good idea to have selected state in this button as Paint and Office apps have?

**Review by 0x5bfa — COMMENTED** [review]
- As a reminder, it might be also better to have a small ‘Expand’ button onto the bottom right of the label as Office apps have.

File: components/Ribbon/samples/Ribbon.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+for all the handled pointer released events. It assumes that if the pointer has been handled
+something reacted to it. It works well for buttons or check boxes but does not work for text
+or combo boxes.
+
+### Priority
+The priority of the group.
+The group with the lower priority will be the first one to be collapsed.
+
+### CollapsedAccessKey
+The access key to access the collapsed button and open the flyout when the group is collapsed.
```
[the remark this answers is not in the document]
Comment (vgromfeld) [reply components/Ribbon/samples/Ribbon.md:55]: When fully visible, there is not access key at the group level. They are defined at the buttons level lige the "Tools "group in Paint:
image

In collapse mode, this access key is the one to open the group's flyout
image

File: components/Ribbon/samples/RibbonCustomSample.xaml [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                <Grid>
+                    <Grid.ColumnDefinitions>
+                        <ColumnDefinition Width="Auto" />
+                        <ColumnDefinition Width="Auto" />
+                        <ColumnDefinition Width="Auto" />
+                    </Grid.ColumnDefinitions>
+                    <Grid.RowDefinitions>
+                        <RowDefinition Height="*" />
+                        <RowDefinition Height="*" />
+                    </Grid.RowDefinitions>
```
[the remark this answers is not in the document]
Comment (vgromfeld) [reply components/Ribbon/samples/RibbonCustomSample.xaml:49]: Indeed, it could be interesting to have a dedicated panel here to avoid repeating the `Grid` and all its setup. 
We can probably add a `MaxLinesOrColumns` property to the `WrapPanel`.

File: components/Ribbon/src/RibbonGroup.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace CommunityToolkit.WinUI.Controls;
+
+/// <summary>
+/// A basic group displayed in a <see cref="Ribbon"/> control.
+/// It adds a <see cref="Label"/> to the wrapped <see cref="Content"/>.
+/// </summary>
+[ContentProperty(Name = nameof(Content))]
+public partial class RibbonGroup : Control
```
[the remark this answers is not in the document]
Comment (vgromfeld) [reply components/Ribbon/src/RibbonGroup.cs:12]: The reparenting does not behave well when inheriting from a `ContentControl`. I'll keep `Control`

File: components/Ribbon/src/RibbonStyle.xaml (hunk 3 above)
Comment (vgromfeld) [reply components/Ribbon/src/RibbonStyle.xaml:112]: This is indeed out of scope of this PR. It can easily be added later.

File: components/Ribbon/samples/RibbonCustomSample.xaml [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+      mc:Ignorable="d">
+
+    <!--
+        TODO: The generic.xaml file from the code project is not imported when the sample is loaded in the gallery app.
+        As a mitigation, we are manually importing the style dictionary here.
+    -->
+    <Page.Resources>
+        <ResourceDictionary Source="ms-appx:///CommunityToolkit.WinUI.Controls.Ribbon/RibbonStyle.xaml" />
+    </Page.Resources>
```
[the remark this answers is not in the document]
Comment (Arlodotexe) [reply components/Ribbon/samples/RibbonCustomSample.xaml:17]: The `Generic.xaml` file is present and working for the `src` project, but the `Generic.xaml` file isn't present in the `samples` project. It needs to be added.

[the remark this answers is not in the document]
Comment (Arlodotexe) [reply components/Ribbon/samples/RibbonCustomSample.xaml:17]: Tried to implement these changes myself, didn't seem to help 🤔
No other component is using a `Generic.xaml` file in the sample project, it seems like something else is happening here.

[the remark this answers is not in the document]
Comment (Arlodotexe) [reply components/Ribbon/samples/RibbonCustomSample.xaml:17]: Interesting, all the existing xaml files (resource dictionaries, controls, etc) are backed by a cs file, except for this new one. We might be auto-importing backed xaml/cs files somewhere in our tooling, but not the xaml files alone.

**Review by michael-hawker — APPROVED** [review]

File: components/Ribbon/samples/RibbonCustomSample.xaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (michael-hawker) [reply components/Ribbon/samples/RibbonCustomSample.xaml:17]: Let's see how it works in a package and we can do a separate fix, if needed. Going to squash and merge this one.

**Comment by Tartarusome** [comment]
May I ask how you made the reveal highlight effect? I checked `RibbonStyle.xaml`, but I found nothing.
image

**Comment by vgromfeld** [comment]
@Tartarusome The reveal effect is on the `GridViewItem`s from the `GridView` used in the "Advanced" group in the sample. Since there is no custom style on those `GridViewItem`, they are picking the system default style.

Note that the reveal effect is no longer part of Windows 11 style.

**Comment by Tartarusome** [comment]
> @Tartarusome The reveal effect is on the `GridViewItem`s from the `GridView` used in the "Advanced" group in the sample. Since there is no custom style on those `GridViewItem`, they are picking the system default style.
> 
> Note that the reveal effect is no longer part of Windows 11 style.

But I truely found this animation in the gallery app on windows11 24h2. How didn't it fallback to default style?
